### PR TITLE
Aligning Workers Integration with EFS and DB

### PR DESCRIPTION
**Pull request by CMCDragonkai**
### Description

The integration of `js-workers` wasn't entirely clean.

The integration style should be following how EFS is integration js-workers.

So there's a bunch of things to remove.

Prior details: https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205/#note_712975388

### Tasks

1. [x] - Remove `src/workers/WorkerManager.test.ts`, `src/workers/errors.ts` and `tests/workers/WorkerManager.test.ts`
2. [x] - Edit `src/workers/polykeyWorkerModule.ts` to remove `encryptWithKey` and `decryptWithKey`
3. ~~[ ] - ArrayBuffer refactoring for `src/workers/polykeyWorkerModule.ts`~~ - held off after overall CLI/API review
4. [x] - Ensure that `workerManager` can be dynamically set for `VaultManager` similar to `KeyManager`
5. [x] - Change `tests/polykeyWorkerModule.test.ts` to` tests/polykeyWorker.test.ts` and update all test structure to match EFS
6. [x] - `PolykeyAgent` and `KeyManager.test.ts` requires updates to how they construct the `PolykeyWorkerManagerInterface`, this can be turned into a utility function if necessary for the default worker

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [ ] Sanity check the final build

### CODE REVIEW

**Comment by joshuakarp** [comment]
As per https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205#note_712994114 and https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205#note_711077414, I'm assuming that the integration of the `workerManager` into `KeyManager` (and therefore the `EFS`) will be handled here.

**Comment by CMCDragonkai** [comment]
Should centralise all WorkerManager injected tests, all other tests should use no worker manager, explicit `null`.

Except process tests, but that's a different beast to be tackled in #264.

**Comment by CMCDragonkai** [comment]
@tegefaulkes please update this issue task list as you go along.

**Comment by tegefaulkes** [comment]
Do we want the `workerManager` type to be `PolykeyWorkerManagerInterface` or should we rename it to `WorkerManager`?

**Comment by CMCDragonkai** [comment]
The `src/workers/utils.ts` `createWorkerManager` needs to be asynchronous.

The `createWorkerManager` call is asynchronous. So the full signature is:

```ts
async function createWorkerManager({
  cores,
  logger,
}: {
  cores?: number;
  logger?: Logger;
}): Promise<PolykeyWorkerManagerInterface> {
  return await WorkerManager.createWorkerManager<PolykeyWorkerModule>({
    workerFactory: () => spawn(new Worker('./polykeyWorker')),
    cores,
    logger,
  });
}
```

**Comment by CMCDragonkai** [comment]
Since the utility function is there, I'm going to move the types in there too from `src/types.ts`.

**Comment by CMCDragonkai** [comment]
@tegefaulkes in the future, beware to use `import type` for things that are types.

**Comment by CMCDragonkai** [comment]
The `workerManager` should not be needed by the notifications domain. I think that was vestigial code left in @emmacasolin.

Only 3 places are using worker manager: `keys` and `vaults` and the DB imported by `js-db`.

**Comment by CMCDragonkai** [comment]
The imports in `PolykeyAgent` are quite crazy. I'd prefer these to be imported as like `workersUtils` and `networkUtils`. That will make it clearer later, but I'll leave that to later piece of work.

**Comment by CMCDragonkai** [comment]
Will leave this `ArrayBuffer refactoring for src/workers/polykeyWorkerModule.ts` till later as part of the CLI/API review #249 and #268 as those functions may or may not be used, and we may have some general crypto refactoring in #270.

**Comment by CMCDragonkai** [comment]
@tegefaulkes in the future try to use qualified names:

```
import * as workersUtils from '@/workers/utils';
```

When importing utilities from a different domain boundary it's better to always use the qualified import because it will allow the code to be understood when posted as snippets.

**Comment by CMCDragonkai** [comment]
Noticed no integration tests of using `VaultManager` with `WorkerManager` together atm, this is because there wasn't even the ability to `setWorkerManager` and `unsetWorkerManager` until now. Will push this to another issue.

**Comment by CMCDragonkai** [comment]
Some full test failures:

```
 FAIL  tests/nodes/NodeGraph.test.ts (5.352 s)
  NodeGraph
    ✓ finds correct node address (12 ms)
    ✓ unable to find node address (6 ms)
    ✓ adds a single node into a bucket (5 ms)
    ✕ adds multiple nodes into the same bucket (21 ms)
    ✓ adds a single node into different buckets (8 ms)
    ✓ deletes a single node (and removes bucket) (7 ms)
    ✕ deletes a single node (and retains remainder of bucket) (20 ms)
    ✓ enforces k-bucket size, removing least active node when a new node is discovered (91 ms)
    ✓ retrieves all buckets (in expected lexicographic order) (16 ms)
    ✓ refreshes buckets (1710 ms)
    ✓ finds a single closest node (4 ms)
    ✓ finds 3 closest nodes (10 ms)
    ✓ finds the 20 closest nodes (98 ms)
    ✓ updates node (5 ms)

  ● NodeGraph › adds multiple nodes into the same bucket

    expect(received).toEqual(expected) // deep equality

    Expected: {"address": {"ip": "6.6.6.6", "port": 6666}, "lastUpdated": Any<Date>}
    Received: undefined

      222 |         lastUpdated: expect.any(Date),
      223 |       });
    > 224 |       expect(bucket[newNode3Id]).toEqual({
          |                                  ^
      225 |         address: { ip: '6.6.6.6', port: 6666 },
      226 |         lastUpdated: expect.any(Date),
      227 |       });

      at Object.<anonymous> (tests/nodes/NodeGraph.test.ts:224:34)

  ● NodeGraph › deletes a single node (and retains remainder of bucket)

    expect(received).toEqual(expected) // deep equality

    Expected: {"address": {"ip": "6.6.6.6", "port": 6666}, "lastUpdated": Any<Date>}
    Received: undefined

      305 |         lastUpdated: expect.any(Date),
      306 |       });
    > 307 |       expect(bucket[newNode3Id]).toEqual({
          |                                  ^
      308 |         address: { ip: '6.6.6.6', port: 6666 },
      309 |         lastUpdated: expect.any(Date),
      310 |       });

      at Object.<anonymous> (tests/nodes/NodeGraph.test.ts:307:34)
```

File: tests/bin/agent.test.ts [hunk 1]
```diff
@@ -162,7 +162,7 @@ describe('CLI agent', () => {
 
           //Checking that the lockfile was removed.
           await poll(global.polykeyStartupTimeout * 2, async () => {
-            const files = await fs.promises.readdir(foregroundNodePath);
+            const files = await fs.promises.readdir(backgroundNodePath);
```
Comment (CMCDragonkai) [inline tests/bin/agent.test.ts:165]: @tegefaulkes why was this changed?

Comment (CMCDragonkai) [reply tests/bin/agent.test.ts:165]: Oh I see this is a stealth fix. Everywhere else around this test is using `backgroundNodePath`. @tegefaulkes always mention any stealth fixes so they aren't stealth.

**Comment by CMCDragonkai** [comment]
The test failure in `tests/nodes/NodeGraph.test.ts` is a regression in this branch. It is passing in master when ran with `npm test -- tests/nodes`, but in this branch it fails. The `nodes` domain doesn't have any usage of `WorkerManager` directly.

Running directly `npm test -- tests/nodes/NodeGraph.test.ts` does work however.

**Comment by CMCDragonkai** [comment]
@joshuakarp I noticed there's a `FIXME` that says that the tests fail randomly. Is this related to this problem? And is there an issue for resolving these random failures?

```
// FIXME, some of these tests fail randomly.
describe('NodeGraph', () => {
```

**Comment by CMCDragonkai** [comment]
The test failures don't seem to be related to this branch's changes. Using `npm test -- tests/nodes --runInBand` works. These tests are concerning given that they take humongous time relative to other tests and they have random failures. I've commented about this in #264 to be resolved afterwards.

**Comment by joshuakarp** [comment]
That `FIXME` tag was added by @tegefaulkes, and it was in the same commit as the async-init patterns being integrated into js-polykey (a month ago). What was the the original reason for it @tegefaulkes?

This might be related to https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205#note_695753874. I have some nodes test utils that generate node IDs for specific bucket indexes, and they may be causing random test failures (where the util function isn't creating a node ID for the expected bucket). Will create an issue for it.

**Comment by tegefaulkes** [comment]
I think that's what it was related to. there are quite a few TODOs and FIXMEs still in the code that should be reviewed and likely removed in most cases. I can make an issue for it and get around to it.

**Comment by joshuakarp** [comment]
All good - already made an issue for this specific `NodeGraph` test failure: https://github.com/MatrixAI/js-polykey/issues/274

Re the TODOs and FIXMEs, there's another comment here regarding this in case you haven't seen it @tegefaulkes https://gitlab.com/MatrixAI/Employees/matrix-team/-/issues/8#note_714211497
